app/src/main/java/com/epfl/drawyourpath/database/FirebaseDatabase.kt

```diff
@@ -0,0 +1,601 @@
+    ): CompletableFuture<Unit> {
+        return addUserIdToFriendList(userId, targetFriend).thenApply {
+            // add the currentUser to the friend list of the user with userId
+            addUserIdToFriendList(targetFriend, userId)
+        }
+    }
+
+    override fun removeFriend(userId: String, targetFriend: String): CompletableFuture<Unit> {
+        return removeUserIdToFriendList(userId, targetFriend).thenApply {
+            // remove the current userId to the friendlist of the user with userId
+            removeUserIdToFriendList(targetFriend, userId)
+        }
+    }
+
+    override fun addRunToHistory(userId: String, run: Run): CompletableFuture<Unit> {
+        // create the field for the new path, the key is the start time
+        return setData(
+            userId,
+            hashMapOf(
+                "${FirebaseKeys.RUN_HISTORY}/${run.getStartTime()}" to run,
+            ),
+        )
+    }
+
+    override fun removeRunFromHistory(userId: String, run: Run): CompletableFuture<Unit> {
+        val future = CompletableFuture<Unit>()
+
+        userRoot(userId).child(FirebaseKeys.RUN_HISTORY).child(run.getStartTime().toString())
+            .removeValue()
+            .addOnSuccessListener { future.complete(Unit) }
+            .addOnFailureListener { future.completeExceptionally(it) }
+
+        return future
+    }
+
+    override fun addDailyGoal(userId: String, dailyGoal: DailyGoal): CompletableFuture<Unit> {
+        val future = CompletableFuture<Unit>()
+
+        // This is how the data is store in the database:
+        // dailyGoals = {
+        //    date = {
+        //        expectedDistance = Value
+        //        expectedTime = Value
+        //        expectedNbOfPaths = Value
+        //        obtainedDistance = Value
+        //        obtainedActivityTime = Value
+        //        obtainedNbOfPaths = Value
+        //    }
+        //    date = {.....}
+        // }
+
+        // transform daily goal to data(except the date)
+        val dailyGoalData = transformDailyGoals(dailyGoal)
+
+        // add the daily goal or update it if it already exist with the new data at this date
+        userRoot(userId).child(FirebaseKeys.DAILY_GOALS)
+            .child(dailyGoal.date.toEpochDay().toString())
+            .updateChildren(dailyGoalData)
+            .addOnSuccessListener { future.complete(Unit) }
+            .addOnFailureListener { future.completeExceptionally(it) }
+
+        return future
+    }
+
+    override fun updateUserAchievements(
+        userId: String,
+        distanceDrawing: Double,
+        activityTimeDrawing: Double,
+    ): CompletableFuture<Unit> {
+        // TODO: Rewrite this function so it can scale as we add achievements.
+        //       Current implementation is too restrictive.
+        /*
+        This is how the achievements are store in the firebase:
+        Users{
+            userId{
+                username: Value
+                ....
```

Comment:
This function need to be rewritten to be scalable, but it is out of scope for the PR. What should we do with it?

Reply:
The achievements are already used in the cash of the run to track the achievements of the user by @HoLuDyEr. We will also used them for the trophies and to track the user performance. Since they are implemented and tested I think it is better to keep them.